File: runSimulation.m
function [o1,o2] = runSimulation(params)
    
    global I_saved t_saved %#ok<GVMISå>
    I_saved = [];
    t_saved = [];

    rng(1)

    % ------------------------
    % Unpack parameters
    % ------------------------
    n = params.n;
    To = params.To;
    tspan = params.tspan;

    Lx = params.Lx; % Size of material
    Ly = params.Ly; % Size of material
    electrode_width = params.electrode_width;
    electrode_height = params.electrode_height;

    % ------------------------
    % Randomize initial particle positions
    % ------------------------
    xo = 10.*rand(n,1); 
    yo = electrode_height.*rand(n,1) - 0.5*electrode_height;

    % ------------------------
    % Zero all initial particle velocities
    % ------------------------
    vox = zeros(n,1);
    voy = zeros(n,1);

    % ------------------------
    % Initial config vector
    % ------------------------
    initial = [xo.',vox.',yo.',voy.',To];

    % ------------------------
    % Run ode simulation of the differentiable force function
    % ------------------------
    tic;
        disp("Starting the solver");
        tic;
        global tindex V
        V = params.V; % Set the initial voltage
        tindex = 1; % Initialize the time index for the simulation
        [t,states]=ode45(@(t,states) calculateForces(t, states, params), tspan, initial);
    toc;

    % TODO: Go over this
    % ------------------------
    
    % A = [t,states];
    % t = A(:,1);               % First column is time
    % states = A(:,2:end);      % Remaining columns are the states

    X_pos = states(:,1:n);      % First n columns are x-positions
    Y_pos = states(:,2*n+1:3*n);  % Third group of n columns are y-positions
    
    %produce 2D plot of particle positions, save as gif
    fig = figure(4);
    h2 = tiledlayout(2,2); 
    ax1 = nexttile([1 2]);

    % Set axis limits and labels once
    xlim(ax1, [-electrode_width - 5, Lx+electrode_width + 5]);
    ylim(ax1, [-electrode_height/2 - 5, electrode_height/2 + 5]);
    xlabel(ax1, 'x-position');
    ylabel(ax1, 'y-position');

    hold(ax1, 'on')  % Keep everything on the same axes

    % Draw patch background once
    patch(ax1, ...
        [-electrode_width, 0, 0, -electrode_width], ...
        [-electrode_height/2, -electrode_height/2, electrode_height/2, electrode_height/2], ...
        [1.0, 0.8431, 0.0], ...
        'EdgeColor', 'none', ...
        'FaceAlpha', 0.3);  % Optional transparency
    
    patch(ax1, ...
        [Lx+electrode_width, Lx, Lx, Lx+electrode_width], ...
        [-electrode_height/2, -electrode_height/2, electrode_height/2, electrode_height/2], ...
        [1.0, 0.8431, 0.0], ...
        'EdgeColor', 'none', ...
        'FaceAlpha', 0.3);  % Optional transparency

    % Plot static pinning sites
    % plot(ax1, params.x_pin, params.y_pin, 'rx', 'MarkerSize', 6, 'LineWidth', 1.5);
    w_pin = params.w_pin;
    R_pin = params.R_pin;

    x = linspace(0, Lx, 100);
    y = linspace(-Ly/2, Ly/2, 100);
    [X, Y] = meshgrid(x, y);
    U = (w_pin / 2) * (sin((2 * pi * (X + Y)) / R_pin) + cos((2 * pi * (X - Y)) / R_pin));
    
    % Transparent surface at z = 0
    surf(ax1, X, Y, zeros(size(U)), U, ...
        'EdgeColor', 'none', 'FaceAlpha', 0.3);  % Apply transparency

    colormap(ax1, 'pink')


    % Initialize handle for particle plot
    hParticles = plot(ax1, X_pos(1,1:n), Y_pos(1,1:n), 'b.');

    for idx=1:length(t)

        % Update particle positions
        set(hParticles, 'XData', X_pos(idx,1:n), 'YData', Y_pos(idx,1:n));

        % Update legend or annotation
        legend(ax1, num2str(t(idx)));

        drawnow

        % Save frame for GIF
        frame = getframe(fig);
        im{idx} = frame2im(frame);

    end

    filename = sprintf('dendrite_growthNEW%d.gif', i);
    for idx = 1:length(t)
        [A2,map] = rgb2ind(im{idx},256);
        if idx == 1
            imwrite(A2,map,filename,'gif','LoopCount',Inf,'DelayTime',0.1);
        else
            imwrite(A2,map,filename,'gif','WriteMode','append','DelayTime',0.1);
        end
    end

    end

File: runSimulation.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import PillowWriter, FuncAnimation
from scipy.integrate import solve_ivp
#import imageio

from calculateForces import calculate_forces


def run_simulation(params):
    """
    Translated from MATLAB runSimulation.m
    Runs the full dendritic growth simulation with visualization and GIF output.
    """

    # Initialize globals (same as MATLAB)
    from calculateForces import I_saved, t_saved
    I_saved.clear()
    t_saved.clear()

    np.random.seed(1)

    # ------------------------
    # Unpack parameters
    # ------------------------
    n = params["n"]
    To = params["T_0"]
    tspan = params["tspan"]

    Lx = params["L_x"]
    Ly = params["L_y"]
    electrode_width = params["electrode_width"]
    electrode_height = params["electrode_height"]

    # ------------------------
    # Randomize initial particle positions
    # ------------------------
    xo = 10 * np.random.rand(n)
    yo = electrode_height * np.random.rand(n) - 0.5 * electrode_height

    # ------------------------
    # Zero all initial particle velocities
    # ------------------------
    vox = np.zeros(n)
    voy = np.zeros(n)

    # ------------------------
    # Initial state vector
    # ------------------------
    initial = np.concatenate([xo, vox, yo, voy, [To]]) # To is wrapped in a list so that concatenate resolves

    # ------------------------
    # Run ODE simulation
    # ------------------------
    print("Starting solver...")
    from calculateForces import tindex, V
    tindex = 1
    V = params["V"]

    if not params["BS"]: #use scipy to integrate
        sol = solve_ivp(
            fun=lambda t, y: calculate_forces(t, y, params),
            t_span=(tspan[0], tspan[-1]),
            y0=initial,
            t_eval=tspan, # recall that tspan is an np.arange array - gives us the integral at evenly spaced intervals of dt. Though the solver may take smaller steps to reduce error.
            method=params['scipy_tag'],
            vectorized=False,
            rtol=1e-6,
            atol=1e-9,
        )

        t = sol.t # times of the solution
        states = sol.y.T # the state vector at each time - should be an array; we transpose so that each row is a time and the columns are the states
    else: #Use Bulirsch-Stoer to integrate
        return

    X_pos = states[:, 0:n] # all times, colums 0 to n-1 of the state vector
    Y_pos = states[:, (2*n):(3*n)] # all times, colums 2n-1 to 3n-1 of the state vector

    # ------------------------
    # Plot setupz
    # ------------------------
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(8, 8))
    plt.tight_layout(pad=3)

    # ---- Top plot: particle evolution ----
    ax1.set_xlim([-electrode_width - 5, Lx + electrode_width + 5])
    ax1.set_ylim([-electrode_height / 2 - 5, electrode_height / 2 + 5])
    ax1.set_xlabel("x-position")
    ax1.set_ylabel("y-position")
    ax1.set_title("Dendritic Growth Simulation")

    # Draw electrodes
    ax1.fill_betweenx(
        [-electrode_height / 2, electrode_height / 2],
        -electrode_width,
        0,
        color=(1.0, 0.8431, 0.0),
        alpha=0.3,
    )
    ax1.fill_betweenx(
        [-electrode_height / 2, electrode_height / 2],
        Lx,
        Lx + electrode_width,
        color=(1.0, 0.8431, 0.0),
        alpha=0.3,
    )

    # ---- Pinning potential background ----
    x = np.linspace(0, Lx, 100)
    y = np.linspace(-Ly / 2, Ly / 2, 100)
    X, Y = np.meshgrid(x, y)
    U = np.zeros_like(X)
    for k in range(len(params["w_pin"])):
        dx = X - params["x_pin"][k]
        dy = Y - params["y_pin"][k]
        U += params["w_pin"][k] * np.exp(-(dx**2 + dy**2) / (params["R_pin"][k] ** 2))

    pcm = ax1.pcolormesh(X, Y, U, cmap="pink", shading="auto", alpha=0.3)
    cb = plt.colorbar(pcm, ax=ax1)
    cb.set_label("Pinning Potential U")

    # ---- Initialize particle positions ----
    particles, = ax1.plot(X_pos[0, :], Y_pos[0, :], "b.", markersize=4)

    # ---- Bottom plot: current vs time ----
    ax2.set_xlim([t[0], t[-1]])
    ax2.set_ylim([0, params["num_e"]])
    ax2.set_xlabel("time")
    ax2.set_ylabel("Current I")
    ax2.grid(True)
    current_line, = ax2.plot([], [], "r-", lw=1.5)

    # ------------------------
    # Animation update function
    # ------------------------
    def update(frame):
        particles.set_data(X_pos[frame, :], Y_pos[frame, :])

        from calculateForces import I_saved, t_saved
        if len(t_saved) > 0:
            mask = np.array(t_saved) <= t[frame]
            if np.any(mask):
                current_line.set_data(np.array(t_saved)[mask], np.array(I_saved)[mask])

        ax1.set_title(f"t = {t[frame]:.2f}")
        return particles, current_line

    # ------------------------
    # Create and save animation
    # ------------------------
    anim = FuncAnimation(fig, update, frames=len(t), interval=100, blit=False)

    filename = "dendrite_growth_simulation.gif"
    print(f"Saving GIF to {filename}...")
    writer = PillowWriter(fps=10)
    anim.save(filename, writer=writer)

    plt.close(fig)
    print("Simulation complete. GIF saved.")

    return t, states
